tests/gateway/test_gateway_service.py

```diff
@@ -0,0 +1,357 @@
+import json
+import os
+
+import pytest
+import requests
+import subprocess
+import time
+import yaml
+
+from mlflow import gateway
+from mlflow.exceptions import MlflowException
+from mlflow.gateway import _start_server, _stop_server, _update_server
+from mlflow.gateway.constants import CONF_PATH_ENV_VAR
+from tests.helper_functions import get_safe_port, LOCALHOST
+
+
+@pytest.fixture
+def basic_config_dict():
+    return [
+        {
+            "name": "instruct-gpt4",
+            "type": "llm/v1/instruct",
+            "model": {
+                "name": "gpt-4",
+                "provider": "openai",
+                "config": {
+                    "openai_api_key": "mykey",
+                    "openai_api_base": "https://api.openai.com/",
+                    "openai_api_version": "v1",
+                    "openai_api_type": "completions",
+                },
+            },
+        },
+        {
+            "name": "claude-chat",
+            "type": "llm/v1/chat",
+            "model": {
+                "name": "claude-v1",
+                "provider": "anthropic",
+                "config": {
+                    "anthropic_api_key": "claudekey",
+                },
+            },
+        },
+    ]
+
+
+@pytest.fixture
+def update_config_dict():
+    return [
+        {
+            "name": "claude-chat",
+            "type": "llm/v1/chat",
+            "model": {
+                "name": "claude-v1",
+                "provider": "anthropic",
+                "config": {
+                    "anthropic_api_key_env_var": "MY_ANTHROPIC_KEY",
+                },
+            },
+        },
+    ]
+
+
+@pytest.fixture
+def invalid_config_dict():
+    return [
+        {
+            "invalid_name": "invalid",
+            "type": "llm/v1/chat",
+            "model": {"invalidkey": "invalid", "invalid_provider": "invalid"},
+        }
+    ]
+
+
+def store_conf(path, name, conf):
+    conf_path = path.joinpath(name)
+    conf_path.write_text(yaml.safe_dump(conf))
+    return conf_path
+
+
+def wait(delay: int = 2) -> None:
+    time.sleep(delay)
+
+
+def get_static_endpoint_response(endpoint, host, port):
+    response = requests.get(f"http://{host}:{port}/{endpoint}")
+    assert response.status_code == 200
+    return response.json()
+
+
+def get_dynamic_route_request(route, data, host, port):
+    response = requests.post(f"http://{host}:{port}/{route}", json=data)
+    assert response.status_code == 200
+    return response.json()
+
+
+def get_invalid_route_request(route, route_type, host, port, error_code, data=None):
+    response = requests.request(method=route_type, url=f"http://{host}:{port}/{route}", json=data)
+    assert response.status_code == error_code
+    return response.text
+
+
+def get_test_client(config, path):
+    from mlflow.gateway.gateway_app import app, _add_routes
+    from mlflow.gateway.handlers import _load_route_config
+    from fastapi.testclient import TestClient
```

Comment:
Can we create a function like `create_app` that takes a config, and creates app from it?

Reply:
I really like this idea :) Working on it!